File: packages/lib-metrics/src/money.ts
// @paradigm: sql
// CF-C2-SUBUNIT-1: Money carries subunit_multiplier — the conversion reads it, never hardcodes 100.
// v1 internal contract: field names are stable; Child 4 imports unchanged.
// Forward note: proto Money message (Child 4) must mirror int64 minor_units / string currency_code / int32 subunit_multiplier.

import { subunitMultiplier } from './subunits.js';

/** Canonical money value object. Always integer minor units (e.g. paise for INR). */
export interface Money {
  /** Integer minor units (e.g. paise). Never float. */
  readonly minorUnits: bigint;
  /** ISO 4217 alpha-3, e.g. "INR". */
  readonly currencyCode: string;
  /**
   * Subunit exponent: 100 (INR/AED/SAR/USD/…), 1000 (KWD/BHD), 1 (JPY).
   * Resolved once at construction from the currency code.
   * CF-C2-SUBUNIT-1: the conversion reads this field, never hardcodes 100.
   */
  readonly subunitMultiplier: number;
}

/**
 * Construct a Money value object.
 * Resolves subunitMultiplier from the currencyCode via the lookup table.
 * CF-C2-SUBUNIT-1 — the lookup, not a hardcoded 100.
 */
export function makeMoney(minorUnits: bigint, currencyCode: string): Money {
  return Object.freeze({
    minorUnits,
    currencyCode: currencyCode.toUpperCase(),
    subunitMultiplier: subunitMultiplier(currencyCode),
  });
}

File: pylibs/brain_metrics/brain_metrics/money.py
"""
money.py — Canonical Money value object.

@paradigm: sql (deterministic value type, zero LLM)
Justified: Money is a pure immutable value object. minor_units is always an
integer (BIGINT/Int64); no float anywhere in this type. CF-C2-PRIMITIVE-1.

Byte-identity pair with packages/lib-metrics/src/money.ts.
"""

from __future__ import annotations

from dataclasses import dataclass

from brain_metrics.subunits import subunit_multiplier as _subunit_multiplier


@dataclass(frozen=True)
class Money:
    """Canonical Brain money representation.

    Fields:
        minor_units: integer minor units (paise for INR, fils for KWD, …).
            Always an int (BIGINT/Int64 semantics). Negative values are valid
            (refunds, credits). Never a float.
        currency_code: ISO 4217 alpha-3 code, e.g. "INR".
        subunit_multiplier: e.g. 100 for INR/AED/SAR, 1000 for KWD/BHD,
            1 for JPY. Stored on the object so consumers never hardcode 100.
            CF-C2-SUBUNIT-1.
    """

    minor_units: int
    currency_code: str
    subunit_multiplier: int


def make_money(minor_units: int, currency_code: str) -> Money:
    """Construct a Money value, resolving subunit_multiplier from currency_code.

    Args:
        minor_units: integer minor units (e.g. 123456 for ₹1234.56).
        currency_code: ISO 4217 alpha-3 (e.g. "INR").

    Returns:
        Money: frozen dataclass instance.
    """
    return Money(
        minor_units=int(minor_units),
        currency_code=currency_code.upper(),
        subunit_multiplier=_subunit_multiplier(currency_code),
    )
